Library/Homebrew/cmd/test-bot.rb

```diff
@@ -944,11 +953,22 @@ def test_bot
 
           if step.has_output?
             # Remove invalid XML CData characters from step output.
-            output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            if ruby_has_encoding?
+              # This is the regex for valid XML chars, but only works in Ruby 2.0+
+              # /[\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/
+              # For 1.9 compatibility, use the inverse of that, which stays under \u10000
+              invalid_xml_pat = /[\x00-\x08\x0B\x0C\x0E-\x1F\uD800-\uDFFF\uFFFE\uFFFF]/
+              output = step.output.gsub(invalid_xml_pat, "\uFFFD")
+            else
+              output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            end
 
             if output.bytesize > BYTES_IN_1_MEGABYTE
-              output = "truncated output to 1MB:\n" \
-                + output.slice(-BYTES_IN_1_MEGABYTE, BYTES_IN_1_MEGABYTE)
+              slice_start = [0, output.length - BYTES_IN_1_MEGABYTE].max
+              if slice_start > 0
+                output = "truncated output to ~1MB:\n" \
+                  + output.slice(slice_start, output.length - slice_start)
+              end
```

Comment:
I'm afraid my previous comment about `length != bytesize` (unless we have pure ASCII) was lost. Let's assume we have a string `x` with `x.length == 1_500_000` and `x.bytesize == 1_800_000` (assuming 20% of code units need two bytes in UTF-8 encoding). If I'm reading the code correctly, `slice` will operate on characters, start at character 451,424, and copy 1,048,576 characters. Still assuming the overhead from multi-byte code units, this will result in 1,258,291 bytes of output (significantly more than the desired 1 MB).

Admittedly, all of that is somewhat constructed and 20% is a big percentage given the output we typically capture. Still could be relevant depending on how sharp this 1 MB limit is.

An easy workaround could be: Truncate first while the encoding is still `BINARY` and convert to UTF-8 (with filtering/fixing the output) afterwards.